// rclnodejs/generated/interfaces/interfaces__action__Fibonacci.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class interfaces__action__Fibonacci {
  static get Goal() {
    return require('./interfaces__action__Fibonacci_Goal.js');
  }
  static get Result() {
    return require('./interfaces__action__Fibonacci_Result.js');
  }
  static get Feedback() {
    return require('./interfaces__action__Fibonacci_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./interfaces__action__Fibonacci_SendGoal.js');
      },
      get GetResultService() {
        return require('./interfaces__action__Fibonacci_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./interfaces__action__Fibonacci_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'interfaces', subFolder: 'action', interfaceName: 'Fibonacci'};
  }
}
module.exports = interfaces__action__Fibonacci;
