// rclnodejs/generated/interfaces/interfaces__msg__PcsYk.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PcsYkRefStruct = StructType({
  start_stop_ctrl: primitiveTypes.bool,
});
const PcsYkRefArray = ArrayType(PcsYkRefStruct);
const PcsYkRefStructArray = StructType({
  data: PcsYkRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PcsYkWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PcsYkRefStruct(other._refObject.toObject());
      this._start_stop_ctrlIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PcsYkRefStruct();
    this._start_stop_ctrlIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PcsYkWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PcsYkArrayWrapper;
  }
  static get ArrayType() {
    return PcsYkArrayWrapper;
  }
  static get refObjectArrayType() {
    return PcsYkRefStructArray
  }
  static get refObjectType() {
    return PcsYkRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._start_stop_ctrlIntialized) {
        throw new TypeError('Invalid argument: start_stop_ctrl in PcsYk');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._start_stop_ctrlIntialized = true;
    this._refObject.start_stop_ctrl = refObject.start_stop_ctrl;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PcsYkWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'interfaces', subFolder: 'msg', interfaceName: 'PcsYk'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get start_stop_ctrl() {
    if (!this._start_stop_ctrlIntialized) {
      return undefined;
    }
    return this._refObject.start_stop_ctrl;
  }
  set start_stop_ctrl(value) {
    this._start_stop_ctrlIntialized = true;
    this._refObject.start_stop_ctrl = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PcsYkRefStruct(refObject.toObject());
    this._start_stop_ctrlIntialized = true;
  }
  copy(other) {
    this._refObject = new PcsYkRefStruct(other._refObject.toObject());
    this._start_stop_ctrlIntialized = true;
  }
  static get classType() {
    return PcsYkWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"start_stop_ctrl","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"interfaces","type":"PcsYk","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PcsYk"};
  }
  hasMember(name) {
    let memberNames = ["start_stop_ctrl"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PcsYkArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PcsYkWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PcsYkWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PcsYkArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PcsYkArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PcsYkRefArray(size);
    this._refObject = new PcsYkRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PcsYkWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PcsYkArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PcsYkArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PcsYkWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PcsYkWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PcsYkArrayWrapper;
  }
}
module.exports = PcsYkWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "start_stop_ctrl",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "interfaces",
    "type": "PcsYk",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PcsYk"
}
*/
