// rclnodejs/generated/interfaces/interfaces__msg__PcsYt.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PcsYtRefStruct = StructType({
  op_mode_ctrl: primitiveTypes.uint32,
  dc_voltage_ctrl: primitiveTypes.float32,
  active_power_ctrl: primitiveTypes.float32,
  reactive_power_ctrl: primitiveTypes.float32,
});
const PcsYtRefArray = ArrayType(PcsYtRefStruct);
const PcsYtRefStructArray = StructType({
  data: PcsYtRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PcsYtWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PcsYtRefStruct(other._refObject.toObject());
      this._op_mode_ctrlIntialized = true;
      this._dc_voltage_ctrlIntialized = true;
      this._active_power_ctrlIntialized = true;
      this._reactive_power_ctrlIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PcsYtRefStruct();
    this._op_mode_ctrlIntialized = false;
    this._dc_voltage_ctrlIntialized = false;
    this._active_power_ctrlIntialized = false;
    this._reactive_power_ctrlIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PcsYtWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PcsYtArrayWrapper;
  }
  static get ArrayType() {
    return PcsYtArrayWrapper;
  }
  static get refObjectArrayType() {
    return PcsYtRefStructArray
  }
  static get refObjectType() {
    return PcsYtRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._op_mode_ctrlIntialized) {
        throw new TypeError('Invalid argument: op_mode_ctrl in PcsYt');
      }
      if (!this._dc_voltage_ctrlIntialized) {
        throw new TypeError('Invalid argument: dc_voltage_ctrl in PcsYt');
      }
      if (!this._active_power_ctrlIntialized) {
        throw new TypeError('Invalid argument: active_power_ctrl in PcsYt');
      }
      if (!this._reactive_power_ctrlIntialized) {
        throw new TypeError('Invalid argument: reactive_power_ctrl in PcsYt');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._op_mode_ctrlIntialized = true;
    this._refObject.op_mode_ctrl = refObject.op_mode_ctrl;
    this._dc_voltage_ctrlIntialized = true;
    this._refObject.dc_voltage_ctrl = refObject.dc_voltage_ctrl;
    this._active_power_ctrlIntialized = true;
    this._refObject.active_power_ctrl = refObject.active_power_ctrl;
    this._reactive_power_ctrlIntialized = true;
    this._refObject.reactive_power_ctrl = refObject.reactive_power_ctrl;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PcsYtWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'interfaces', subFolder: 'msg', interfaceName: 'PcsYt'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get op_mode_ctrl() {
    if (!this._op_mode_ctrlIntialized) {
      return undefined;
    }
    return this._refObject.op_mode_ctrl;
  }
  set op_mode_ctrl(value) {
    this._op_mode_ctrlIntialized = true;
    this._refObject.op_mode_ctrl = value;
  }
  get dc_voltage_ctrl() {
    if (!this._dc_voltage_ctrlIntialized) {
      return undefined;
    }
    return this._refObject.dc_voltage_ctrl;
  }
  set dc_voltage_ctrl(value) {
    this._dc_voltage_ctrlIntialized = true;
    this._refObject.dc_voltage_ctrl = value;
  }
  get active_power_ctrl() {
    if (!this._active_power_ctrlIntialized) {
      return undefined;
    }
    return this._refObject.active_power_ctrl;
  }
  set active_power_ctrl(value) {
    this._active_power_ctrlIntialized = true;
    this._refObject.active_power_ctrl = value;
  }
  get reactive_power_ctrl() {
    if (!this._reactive_power_ctrlIntialized) {
      return undefined;
    }
    return this._refObject.reactive_power_ctrl;
  }
  set reactive_power_ctrl(value) {
    this._reactive_power_ctrlIntialized = true;
    this._refObject.reactive_power_ctrl = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PcsYtRefStruct(refObject.toObject());
    this._op_mode_ctrlIntialized = true;
    this._dc_voltage_ctrlIntialized = true;
    this._active_power_ctrlIntialized = true;
    this._reactive_power_ctrlIntialized = true;
  }
  copy(other) {
    this._refObject = new PcsYtRefStruct(other._refObject.toObject());
    this._op_mode_ctrlIntialized = true;
    this._dc_voltage_ctrlIntialized = true;
    this._active_power_ctrlIntialized = true;
    this._reactive_power_ctrlIntialized = true;
  }
  static get classType() {
    return PcsYtWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"op_mode_ctrl","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dc_voltage_ctrl","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"active_power_ctrl","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reactive_power_ctrl","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"interfaces","type":"PcsYt","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PcsYt"};
  }
  hasMember(name) {
    let memberNames = ["op_mode_ctrl","dc_voltage_ctrl","active_power_ctrl","reactive_power_ctrl"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PcsYtArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PcsYtWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PcsYtWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PcsYtArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PcsYtArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PcsYtRefArray(size);
    this._refObject = new PcsYtRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PcsYtWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PcsYtArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PcsYtArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PcsYtWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PcsYtWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PcsYtArrayWrapper;
  }
}
module.exports = PcsYtWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "op_mode_ctrl",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dc_voltage_ctrl",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "active_power_ctrl",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reactive_power_ctrl",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "interfaces",
    "type": "PcsYt",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PcsYt"
}
*/
