// rclnodejs/generated/interfaces/interfaces__msg__LeakTestResult.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LeakTestResultRefStruct = StructType({
  leak_rate: primitiveTypes.float64,
  test_pressure: primitiveTypes.float64,
  duration: primitiveTypes.float64,
  status: primitiveTypes.int8,
});
const LeakTestResultRefArray = ArrayType(LeakTestResultRefStruct);
const LeakTestResultRefStructArray = StructType({
  data: LeakTestResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LeakTestResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LeakTestResultRefStruct(other._refObject.toObject());
      this._leak_rateIntialized = true;
      this._test_pressureIntialized = true;
      this._durationIntialized = true;
      this._statusIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LeakTestResultRefStruct();
    this._leak_rateIntialized = false;
    this._test_pressureIntialized = false;
    this._durationIntialized = false;
    this._statusIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LeakTestResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LeakTestResultArrayWrapper;
  }
  static get ArrayType() {
    return LeakTestResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return LeakTestResultRefStructArray
  }
  static get refObjectType() {
    return LeakTestResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._leak_rateIntialized) {
        throw new TypeError('Invalid argument: leak_rate in LeakTestResult');
      }
      if (!this._test_pressureIntialized) {
        throw new TypeError('Invalid argument: test_pressure in LeakTestResult');
      }
      if (!this._durationIntialized) {
        throw new TypeError('Invalid argument: duration in LeakTestResult');
      }
      if (!this._statusIntialized) {
        throw new TypeError('Invalid argument: status in LeakTestResult');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._leak_rateIntialized = true;
    this._refObject.leak_rate = refObject.leak_rate;
    this._test_pressureIntialized = true;
    this._refObject.test_pressure = refObject.test_pressure;
    this._durationIntialized = true;
    this._refObject.duration = refObject.duration;
    this._statusIntialized = true;
    this._refObject.status = refObject.status;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LeakTestResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'interfaces', subFolder: 'msg', interfaceName: 'LeakTestResult'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get leak_rate() {
    if (!this._leak_rateIntialized) {
      return undefined;
    }
    return this._refObject.leak_rate;
  }
  set leak_rate(value) {
    this._leak_rateIntialized = true;
    this._refObject.leak_rate = value;
  }
  get test_pressure() {
    if (!this._test_pressureIntialized) {
      return undefined;
    }
    return this._refObject.test_pressure;
  }
  set test_pressure(value) {
    this._test_pressureIntialized = true;
    this._refObject.test_pressure = value;
  }
  get duration() {
    if (!this._durationIntialized) {
      return undefined;
    }
    return this._refObject.duration;
  }
  set duration(value) {
    this._durationIntialized = true;
    this._refObject.duration = value;
  }
  get status() {
    if (!this._statusIntialized) {
      return undefined;
    }
    return this._refObject.status;
  }
  set status(value) {
    this._statusIntialized = true;
    this._refObject.status = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LeakTestResultRefStruct(refObject.toObject());
    this._leak_rateIntialized = true;
    this._test_pressureIntialized = true;
    this._durationIntialized = true;
    this._statusIntialized = true;
  }
  copy(other) {
    this._refObject = new LeakTestResultRefStruct(other._refObject.toObject());
    this._leak_rateIntialized = true;
    this._test_pressureIntialized = true;
    this._durationIntialized = true;
    this._statusIntialized = true;
  }
  static get classType() {
    return LeakTestResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"leak_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"test_pressure","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"interfaces","type":"LeakTestResult","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LeakTestResult"};
  }
  hasMember(name) {
    let memberNames = ["leak_rate","test_pressure","duration","status"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LeakTestResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LeakTestResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LeakTestResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LeakTestResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LeakTestResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LeakTestResultRefArray(size);
    this._refObject = new LeakTestResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LeakTestResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LeakTestResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LeakTestResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LeakTestResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LeakTestResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LeakTestResultArrayWrapper;
  }
}
module.exports = LeakTestResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "leak_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "test_pressure",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "duration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "interfaces",
    "type": "LeakTestResult",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LeakTestResult"
}
*/
