// rclnodejs/generated/interfaces/interfaces__msg__PcsYx.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PcsYxRefStruct = StructType({
  short_circuit_alarm: primitiveTypes.bool,
  polarity_reversal_alarm: primitiveTypes.bool,
  dc_over_voltage_alarm: primitiveTypes.bool,
  dc_under_voltage_alarm: primitiveTypes.bool,
  over_current_alarm: primitiveTypes.bool,
  over_temperature_alarm: primitiveTypes.bool,
  hardware_failure_alarm: primitiveTypes.bool,
  anti_islandin_alarm: primitiveTypes.bool,
  comm_fail_alarm: primitiveTypes.bool,
  short_circuit_prot: primitiveTypes.bool,
  polarity_reversal_prot: primitiveTypes.bool,
  dc_over_voltage_prot: primitiveTypes.bool,
  dc_under_voltage_prot: primitiveTypes.bool,
  over_current_prot: primitiveTypes.bool,
  over_temperature_prot: primitiveTypes.bool,
  hardware_failure_prot: primitiveTypes.bool,
  anti_islandin_prot: primitiveTypes.bool,
  comm_fail_prot: primitiveTypes.bool,
});
const PcsYxRefArray = ArrayType(PcsYxRefStruct);
const PcsYxRefStructArray = StructType({
  data: PcsYxRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PcsYxWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PcsYxRefStruct(other._refObject.toObject());
      this._short_circuit_alarmIntialized = true;
      this._polarity_reversal_alarmIntialized = true;
      this._dc_over_voltage_alarmIntialized = true;
      this._dc_under_voltage_alarmIntialized = true;
      this._over_current_alarmIntialized = true;
      this._over_temperature_alarmIntialized = true;
      this._hardware_failure_alarmIntialized = true;
      this._anti_islandin_alarmIntialized = true;
      this._comm_fail_alarmIntialized = true;
      this._short_circuit_protIntialized = true;
      this._polarity_reversal_protIntialized = true;
      this._dc_over_voltage_protIntialized = true;
      this._dc_under_voltage_protIntialized = true;
      this._over_current_protIntialized = true;
      this._over_temperature_protIntialized = true;
      this._hardware_failure_protIntialized = true;
      this._anti_islandin_protIntialized = true;
      this._comm_fail_protIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PcsYxRefStruct();
    this._short_circuit_alarmIntialized = false;
    this._polarity_reversal_alarmIntialized = false;
    this._dc_over_voltage_alarmIntialized = false;
    this._dc_under_voltage_alarmIntialized = false;
    this._over_current_alarmIntialized = false;
    this._over_temperature_alarmIntialized = false;
    this._hardware_failure_alarmIntialized = false;
    this._anti_islandin_alarmIntialized = false;
    this._comm_fail_alarmIntialized = false;
    this._short_circuit_protIntialized = false;
    this._polarity_reversal_protIntialized = false;
    this._dc_over_voltage_protIntialized = false;
    this._dc_under_voltage_protIntialized = false;
    this._over_current_protIntialized = false;
    this._over_temperature_protIntialized = false;
    this._hardware_failure_protIntialized = false;
    this._anti_islandin_protIntialized = false;
    this._comm_fail_protIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PcsYxWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PcsYxArrayWrapper;
  }
  static get ArrayType() {
    return PcsYxArrayWrapper;
  }
  static get refObjectArrayType() {
    return PcsYxRefStructArray
  }
  static get refObjectType() {
    return PcsYxRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._short_circuit_alarmIntialized) {
        throw new TypeError('Invalid argument: short_circuit_alarm in PcsYx');
      }
      if (!this._polarity_reversal_alarmIntialized) {
        throw new TypeError('Invalid argument: polarity_reversal_alarm in PcsYx');
      }
      if (!this._dc_over_voltage_alarmIntialized) {
        throw new TypeError('Invalid argument: dc_over_voltage_alarm in PcsYx');
      }
      if (!this._dc_under_voltage_alarmIntialized) {
        throw new TypeError('Invalid argument: dc_under_voltage_alarm in PcsYx');
      }
      if (!this._over_current_alarmIntialized) {
        throw new TypeError('Invalid argument: over_current_alarm in PcsYx');
      }
      if (!this._over_temperature_alarmIntialized) {
        throw new TypeError('Invalid argument: over_temperature_alarm in PcsYx');
      }
      if (!this._hardware_failure_alarmIntialized) {
        throw new TypeError('Invalid argument: hardware_failure_alarm in PcsYx');
      }
      if (!this._anti_islandin_alarmIntialized) {
        throw new TypeError('Invalid argument: anti_islandin_alarm in PcsYx');
      }
      if (!this._comm_fail_alarmIntialized) {
        throw new TypeError('Invalid argument: comm_fail_alarm in PcsYx');
      }
      if (!this._short_circuit_protIntialized) {
        throw new TypeError('Invalid argument: short_circuit_prot in PcsYx');
      }
      if (!this._polarity_reversal_protIntialized) {
        throw new TypeError('Invalid argument: polarity_reversal_prot in PcsYx');
      }
      if (!this._dc_over_voltage_protIntialized) {
        throw new TypeError('Invalid argument: dc_over_voltage_prot in PcsYx');
      }
      if (!this._dc_under_voltage_protIntialized) {
        throw new TypeError('Invalid argument: dc_under_voltage_prot in PcsYx');
      }
      if (!this._over_current_protIntialized) {
        throw new TypeError('Invalid argument: over_current_prot in PcsYx');
      }
      if (!this._over_temperature_protIntialized) {
        throw new TypeError('Invalid argument: over_temperature_prot in PcsYx');
      }
      if (!this._hardware_failure_protIntialized) {
        throw new TypeError('Invalid argument: hardware_failure_prot in PcsYx');
      }
      if (!this._anti_islandin_protIntialized) {
        throw new TypeError('Invalid argument: anti_islandin_prot in PcsYx');
      }
      if (!this._comm_fail_protIntialized) {
        throw new TypeError('Invalid argument: comm_fail_prot in PcsYx');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._short_circuit_alarmIntialized = true;
    this._refObject.short_circuit_alarm = refObject.short_circuit_alarm;
    this._polarity_reversal_alarmIntialized = true;
    this._refObject.polarity_reversal_alarm = refObject.polarity_reversal_alarm;
    this._dc_over_voltage_alarmIntialized = true;
    this._refObject.dc_over_voltage_alarm = refObject.dc_over_voltage_alarm;
    this._dc_under_voltage_alarmIntialized = true;
    this._refObject.dc_under_voltage_alarm = refObject.dc_under_voltage_alarm;
    this._over_current_alarmIntialized = true;
    this._refObject.over_current_alarm = refObject.over_current_alarm;
    this._over_temperature_alarmIntialized = true;
    this._refObject.over_temperature_alarm = refObject.over_temperature_alarm;
    this._hardware_failure_alarmIntialized = true;
    this._refObject.hardware_failure_alarm = refObject.hardware_failure_alarm;
    this._anti_islandin_alarmIntialized = true;
    this._refObject.anti_islandin_alarm = refObject.anti_islandin_alarm;
    this._comm_fail_alarmIntialized = true;
    this._refObject.comm_fail_alarm = refObject.comm_fail_alarm;
    this._short_circuit_protIntialized = true;
    this._refObject.short_circuit_prot = refObject.short_circuit_prot;
    this._polarity_reversal_protIntialized = true;
    this._refObject.polarity_reversal_prot = refObject.polarity_reversal_prot;
    this._dc_over_voltage_protIntialized = true;
    this._refObject.dc_over_voltage_prot = refObject.dc_over_voltage_prot;
    this._dc_under_voltage_protIntialized = true;
    this._refObject.dc_under_voltage_prot = refObject.dc_under_voltage_prot;
    this._over_current_protIntialized = true;
    this._refObject.over_current_prot = refObject.over_current_prot;
    this._over_temperature_protIntialized = true;
    this._refObject.over_temperature_prot = refObject.over_temperature_prot;
    this._hardware_failure_protIntialized = true;
    this._refObject.hardware_failure_prot = refObject.hardware_failure_prot;
    this._anti_islandin_protIntialized = true;
    this._refObject.anti_islandin_prot = refObject.anti_islandin_prot;
    this._comm_fail_protIntialized = true;
    this._refObject.comm_fail_prot = refObject.comm_fail_prot;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PcsYxWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'interfaces', subFolder: 'msg', interfaceName: 'PcsYx'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get short_circuit_alarm() {
    if (!this._short_circuit_alarmIntialized) {
      return undefined;
    }
    return this._refObject.short_circuit_alarm;
  }
  set short_circuit_alarm(value) {
    this._short_circuit_alarmIntialized = true;
    this._refObject.short_circuit_alarm = value;
  }
  get polarity_reversal_alarm() {
    if (!this._polarity_reversal_alarmIntialized) {
      return undefined;
    }
    return this._refObject.polarity_reversal_alarm;
  }
  set polarity_reversal_alarm(value) {
    this._polarity_reversal_alarmIntialized = true;
    this._refObject.polarity_reversal_alarm = value;
  }
  get dc_over_voltage_alarm() {
    if (!this._dc_over_voltage_alarmIntialized) {
      return undefined;
    }
    return this._refObject.dc_over_voltage_alarm;
  }
  set dc_over_voltage_alarm(value) {
    this._dc_over_voltage_alarmIntialized = true;
    this._refObject.dc_over_voltage_alarm = value;
  }
  get dc_under_voltage_alarm() {
    if (!this._dc_under_voltage_alarmIntialized) {
      return undefined;
    }
    return this._refObject.dc_under_voltage_alarm;
  }
  set dc_under_voltage_alarm(value) {
    this._dc_under_voltage_alarmIntialized = true;
    this._refObject.dc_under_voltage_alarm = value;
  }
  get over_current_alarm() {
    if (!this._over_current_alarmIntialized) {
      return undefined;
    }
    return this._refObject.over_current_alarm;
  }
  set over_current_alarm(value) {
    this._over_current_alarmIntialized = true;
    this._refObject.over_current_alarm = value;
  }
  get over_temperature_alarm() {
    if (!this._over_temperature_alarmIntialized) {
      return undefined;
    }
    return this._refObject.over_temperature_alarm;
  }
  set over_temperature_alarm(value) {
    this._over_temperature_alarmIntialized = true;
    this._refObject.over_temperature_alarm = value;
  }
  get hardware_failure_alarm() {
    if (!this._hardware_failure_alarmIntialized) {
      return undefined;
    }
    return this._refObject.hardware_failure_alarm;
  }
  set hardware_failure_alarm(value) {
    this._hardware_failure_alarmIntialized = true;
    this._refObject.hardware_failure_alarm = value;
  }
  get anti_islandin_alarm() {
    if (!this._anti_islandin_alarmIntialized) {
      return undefined;
    }
    return this._refObject.anti_islandin_alarm;
  }
  set anti_islandin_alarm(value) {
    this._anti_islandin_alarmIntialized = true;
    this._refObject.anti_islandin_alarm = value;
  }
  get comm_fail_alarm() {
    if (!this._comm_fail_alarmIntialized) {
      return undefined;
    }
    return this._refObject.comm_fail_alarm;
  }
  set comm_fail_alarm(value) {
    this._comm_fail_alarmIntialized = true;
    this._refObject.comm_fail_alarm = value;
  }
  get short_circuit_prot() {
    if (!this._short_circuit_protIntialized) {
      return undefined;
    }
    return this._refObject.short_circuit_prot;
  }
  set short_circuit_prot(value) {
    this._short_circuit_protIntialized = true;
    this._refObject.short_circuit_prot = value;
  }
  get polarity_reversal_prot() {
    if (!this._polarity_reversal_protIntialized) {
      return undefined;
    }
    return this._refObject.polarity_reversal_prot;
  }
  set polarity_reversal_prot(value) {
    this._polarity_reversal_protIntialized = true;
    this._refObject.polarity_reversal_prot = value;
  }
  get dc_over_voltage_prot() {
    if (!this._dc_over_voltage_protIntialized) {
      return undefined;
    }
    return this._refObject.dc_over_voltage_prot;
  }
  set dc_over_voltage_prot(value) {
    this._dc_over_voltage_protIntialized = true;
    this._refObject.dc_over_voltage_prot = value;
  }
  get dc_under_voltage_prot() {
    if (!this._dc_under_voltage_protIntialized) {
      return undefined;
    }
    return this._refObject.dc_under_voltage_prot;
  }
  set dc_under_voltage_prot(value) {
    this._dc_under_voltage_protIntialized = true;
    this._refObject.dc_under_voltage_prot = value;
  }
  get over_current_prot() {
    if (!this._over_current_protIntialized) {
      return undefined;
    }
    return this._refObject.over_current_prot;
  }
  set over_current_prot(value) {
    this._over_current_protIntialized = true;
    this._refObject.over_current_prot = value;
  }
  get over_temperature_prot() {
    if (!this._over_temperature_protIntialized) {
      return undefined;
    }
    return this._refObject.over_temperature_prot;
  }
  set over_temperature_prot(value) {
    this._over_temperature_protIntialized = true;
    this._refObject.over_temperature_prot = value;
  }
  get hardware_failure_prot() {
    if (!this._hardware_failure_protIntialized) {
      return undefined;
    }
    return this._refObject.hardware_failure_prot;
  }
  set hardware_failure_prot(value) {
    this._hardware_failure_protIntialized = true;
    this._refObject.hardware_failure_prot = value;
  }
  get anti_islandin_prot() {
    if (!this._anti_islandin_protIntialized) {
      return undefined;
    }
    return this._refObject.anti_islandin_prot;
  }
  set anti_islandin_prot(value) {
    this._anti_islandin_protIntialized = true;
    this._refObject.anti_islandin_prot = value;
  }
  get comm_fail_prot() {
    if (!this._comm_fail_protIntialized) {
      return undefined;
    }
    return this._refObject.comm_fail_prot;
  }
  set comm_fail_prot(value) {
    this._comm_fail_protIntialized = true;
    this._refObject.comm_fail_prot = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PcsYxRefStruct(refObject.toObject());
    this._short_circuit_alarmIntialized = true;
    this._polarity_reversal_alarmIntialized = true;
    this._dc_over_voltage_alarmIntialized = true;
    this._dc_under_voltage_alarmIntialized = true;
    this._over_current_alarmIntialized = true;
    this._over_temperature_alarmIntialized = true;
    this._hardware_failure_alarmIntialized = true;
    this._anti_islandin_alarmIntialized = true;
    this._comm_fail_alarmIntialized = true;
    this._short_circuit_protIntialized = true;
    this._polarity_reversal_protIntialized = true;
    this._dc_over_voltage_protIntialized = true;
    this._dc_under_voltage_protIntialized = true;
    this._over_current_protIntialized = true;
    this._over_temperature_protIntialized = true;
    this._hardware_failure_protIntialized = true;
    this._anti_islandin_protIntialized = true;
    this._comm_fail_protIntialized = true;
  }
  copy(other) {
    this._refObject = new PcsYxRefStruct(other._refObject.toObject());
    this._short_circuit_alarmIntialized = true;
    this._polarity_reversal_alarmIntialized = true;
    this._dc_over_voltage_alarmIntialized = true;
    this._dc_under_voltage_alarmIntialized = true;
    this._over_current_alarmIntialized = true;
    this._over_temperature_alarmIntialized = true;
    this._hardware_failure_alarmIntialized = true;
    this._anti_islandin_alarmIntialized = true;
    this._comm_fail_alarmIntialized = true;
    this._short_circuit_protIntialized = true;
    this._polarity_reversal_protIntialized = true;
    this._dc_over_voltage_protIntialized = true;
    this._dc_under_voltage_protIntialized = true;
    this._over_current_protIntialized = true;
    this._over_temperature_protIntialized = true;
    this._hardware_failure_protIntialized = true;
    this._anti_islandin_protIntialized = true;
    this._comm_fail_protIntialized = true;
  }
  static get classType() {
    return PcsYxWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"short_circuit_alarm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"polarity_reversal_alarm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dc_over_voltage_alarm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dc_under_voltage_alarm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"over_current_alarm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"over_temperature_alarm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hardware_failure_alarm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"anti_islandin_alarm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"comm_fail_alarm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"short_circuit_prot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"polarity_reversal_prot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dc_over_voltage_prot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dc_under_voltage_prot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"over_current_prot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"over_temperature_prot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hardware_failure_prot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"anti_islandin_prot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"comm_fail_prot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"interfaces","type":"PcsYx","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PcsYx"};
  }
  hasMember(name) {
    let memberNames = ["short_circuit_alarm","polarity_reversal_alarm","dc_over_voltage_alarm","dc_under_voltage_alarm","over_current_alarm","over_temperature_alarm","hardware_failure_alarm","anti_islandin_alarm","comm_fail_alarm","short_circuit_prot","polarity_reversal_prot","dc_over_voltage_prot","dc_under_voltage_prot","over_current_prot","over_temperature_prot","hardware_failure_prot","anti_islandin_prot","comm_fail_prot"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PcsYxArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PcsYxWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PcsYxWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PcsYxArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PcsYxArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PcsYxRefArray(size);
    this._refObject = new PcsYxRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PcsYxWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PcsYxArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PcsYxArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PcsYxWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PcsYxWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PcsYxArrayWrapper;
  }
}
module.exports = PcsYxWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "short_circuit_alarm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "polarity_reversal_alarm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dc_over_voltage_alarm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dc_under_voltage_alarm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "over_current_alarm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "over_temperature_alarm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hardware_failure_alarm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "anti_islandin_alarm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "comm_fail_alarm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "short_circuit_prot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "polarity_reversal_prot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dc_over_voltage_prot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dc_under_voltage_prot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "over_current_prot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "over_temperature_prot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hardware_failure_prot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "anti_islandin_prot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "comm_fail_prot",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "interfaces",
    "type": "PcsYx",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PcsYx"
}
*/
